### PR TITLE
make CancelConnectAsync thread safe

### PR DESCRIPTION
**Pull request by wfurt**
CancelConnectAsync uses global state which can change during execution of the function. That is clear from the core linked in #42085. We can hit Assert if connect finishes and both _multipleConnect and _currentSocket will be null or we can hit NRE. 

To make the function more thread safe this change captures _multipleConnect to local variable and add check to prevent NRE. 

I did ~ 1000 runs and I did not see any more failures where it it would fail in ~100ish iterations before.

### CODE REVIEW

**Review by geoffkizer — APPROVED** [review]

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs [hunk 1]
```diff
@@ -565,22 +565,18 @@ internal void CancelConnectAsync()
         {
             if (_operating == InProgress && _completedOperation == SocketAsyncOperation.Connect)
             {
-                if (_multipleConnect != null)
+                var multipleConnect = _multipleConnect;
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs:568]: Nit: var => whatever type it is

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs (hunk 1 above)
Comment (wfurt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs]: updated.

**Comment by danmosemsft** [comment]
Is there customer impact that would justify 5.0 port?

**Comment by wfurt** [comment]
I don't think so at the moment. This was found by running our tests in tight loop. There is no customer report so far. Not 5.0 regression AFAIK. Also it was Debug.Assert so it will not be in production code. NRE can be handled if needed as workaround.

**Pull request merged**